p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,230 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file defines the overall structure of a P4 program in our IR, as well
+// as the supported statements and expressions.
+// The structure is tailored to be easy to consume rather than produce.
+// It deviates from the underlying bmv2 JSON format, which is mimiced closely
+// by p4_symbolic/bmv2/bmv2.proto.
+// The bmv2 proto is used to initially parse the JSON format, which is then
+// processed by our tool and transformed into the IR.
+// The IR does not cover all the features of P4 programs.
+// Specifically, we intentionally ignore parsers, de-parsers, stacks and unions,
+// checksum calculations, and extern definitions.
+// Additionally, some types of expressions and statements are unsupported, these
+// are described in the relevant sections of this file.
+// See the bmv2 JSON format reference for more information:
+// https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md
+
+syntax = "proto3";
+
+package p4_symbolic.ir;
+
+import "p4_pdpi/ir.proto";
+import "p4_symbolic/bmv2/bmv2.proto";
+
+// Overall structure of a p4 program.
+message P4Program {
+  // Header definitions, keyed by the header type name (e.g. standard_metadata).
+  map<string, HeaderType> headers = 1;
+  // Action definitions and implementations, keyed by action name.
+  map<string, Action> actions = 2;
+  // Table definitions and implementations, keyed by table name.
+  map<string, Table> tables = 3;
+  // The name of the initial table.
+  string initial_table = 6;
+}
+
+// A header definition. Includes the programmer assigned name,
+// the header id, and the fields in the heade.
+message HeaderType {
+  string name = 1;
+  int32 id = 2;
+  // Maps a field name to its definition.
+  map<string, HeaderField> fields = 3;
+}
+
+// A P4 header type definition.
+// Includes its name, bitwidth (type), and sign flag, and the
+// enclosing header type name that defines this field.
+message HeaderField {
+  string name = 1;
+  int32 bitwidth = 2;
+  bool signed = 3;
+  string header_type = 4;
+}
+
+// Overall action structure, combining definition with implementation.
+message Action {
+  // This contains the name, id, and parameters, among other things.
```

Comment:
Similarly, I don't think in comments for a field you just want to list the fields of that message. Instead, say on a higher level what the field is (or leave it off if you feel that it is clear enough from the name). I would leave it off here. We should probably not say "essentially the action signature", since this is not adding more information, only introducing another name for the same concept, and we generally want to make sure we consistently only use one name for every concept.

Same for action implementation, either say something like "The implementation of this action as a sequence of statements", or leave it off.